testomania/src/main/java/com/earth/testomania/common/networking/NetworkConnectivityObserver.kt

```diff
@@ -0,0 +1,56 @@
+package com.earth.testomania.common.networking
+
+import android.content.Context
+import android.net.*
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Created by Vazhapp on 23.10.2022
+ */
+@Singleton
+class NetworkConnectivityObserver @Inject constructor(
+    @ApplicationContext context: Context,
+) : ConnectivityObserver {
+
+    private var connectivityManager: ConnectivityManager =
+        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+    override fun observe(): Flow<ConnectivityObserver.ConnectionState> {
+        return callbackFlow {
+            val callBack = object : ConnectivityManager.NetworkCallback() {
+                override fun onAvailable(network: Network) {
+                    super.onAvailable(network)
+                    launch { trySend(ConnectivityObserver.ConnectionState.Available) }
+                }
+
+                override fun onLost(network: Network) {
+                    super.onLost(network)
+                    launch { trySend(ConnectivityObserver.ConnectionState.Unavailable) }
+                }
+            }
+            // guess immediately device is connected to the internet or not
+            launch { trySend(checkCurrentNetworkState()) }
+
+            connectivityManager.registerDefaultNetworkCallback(callBack)
+            awaitClose {
+                connectivityManager.unregisterNetworkCallback(callBack)
+            }
+        }.distinctUntilChanged()
```

Comment:
I found this doc for distinctUntilChanged()

```
Returns flow where all subsequent repetitions of the same value are filtered out.
Note that any instance of StateFlow already behaves as if distinctUntilChanged operator is applied to it, 
so applying distinctUntilChanged to a StateFlow has no effect. See StateFlow documentation on Operator Fusion. 
Also, repeated application of distinctUntilChanged operator on any flow has no effect.
```
Do we still need to call this function for flow?